### PR TITLE
Added secondary_indices as a default arg

### LINKED ISSUE #46: Keyword argument for secondary_indices is not used consistently in the code base
We introduced the keyword argument `secondary_indices` to allow writing pipelines to calculate an index on the fly instead of having the user to provide this as an input.
This keyword has been introduced only inconsistently, yet and many pipelines still don't use it. We should enable all writing pipelines to use this keyword argument to eventually deprecate the explicit user input.

### PR DESCRIPTION
**Pull request by aniruddhgoteti**
# Description:

Added a consistant arg `secondary_indices' to all writing pipelines

- [x] Closes #46 
- [x] Tests added / passed
- [x] Passes `./format_code.sh`
- [x] Changelog entry

### CODE REVIEW

File: kartothek/io/dask/bag.py [hunk 1]
```diff
@@ -56,6 +56,7 @@ def read_dataset_as_metapartitions_bag(
     factory=None,
     dispatch_by=None,
     partition_size=None,
+    secondary_indices=None,
```
Comment (fjetter) [inline kartothek/io/dask/bag.py]: This is a read function. No need for the secondary indices keyword

File: kartothek/io/dask/bag.py [hunk 2]
```diff
@@ -239,7 +246,7 @@ def store_bag_as_dataset(
 
 @default_docs
 def build_dataset_indices__bag(
-    store, dataset_uuid, columns, partition_size=None, factory=None
+    store, dataset_uuid, columns, partition_size=None, factory=None, secondary_indices=None,
```
Comment (fjetter) [inline kartothek/io/dask/bag.py]: This pipeline is intended for building indices and the argument `columns` covers already that use case

File: kartothek/io/dask/dataframe.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ def read_dataset_as_ddf(
     dates_as_object=False,
     predicates=None,
     factory=None,
+    secondary_indices=None,
```
Comment (fjetter) [inline kartothek/io/dask/dataframe.py]: Also a read pipeline

File: kartothek/io/dask/delayed.py [hunk 4]
```diff
@@ -156,6 +156,8 @@ def merge_datasets_as_delayed(
     match_how="exact",
     label_merger=None,
     metadata_merger=None,
+    secondary_indices=None,
```
Comment (fjetter) [inline kartothek/io/dask/delayed.py]: This may actually be valid :)

File: kartothek/io/dask/delayed.py [hunk 5]
```diff
@@ -273,6 +278,8 @@ def read_dataset_as_delayed_metapartitions(
     predicates=None,
     factory=None,
     dispatch_by=None,
+    secondary_indices=None,
```
Comment (fjetter) [inline kartothek/io/dask/delayed.py]: again read

**Comment by aniruddhgoteti** [comment]
Also do I need to write entry in the changelog for this?

**Comment by aniruddhgoteti** [comment]
> Why do you require the secondary indices keyword for read pipelines?

Misread only `write` pipelines as `all` pipelines

**Comment by fjetter** [comment]
Can you describe the use case for having index build in read pipelines?

**Comment by fjetter** [comment]
Yes, please also add a changelog entry

**Comment by aniruddhgoteti** [comment]
> Please rebase with the latest master and update the changelog accordingly

Done

File: kartothek/io/eager.py [hunk 6]
```diff
@@ -497,6 +502,9 @@ def store_dataframes_as_dataset(
         store=store, dataset_uuid=dataset_uuid, df_serializer=df_serializer
     )
 
+    if secondary_indices:
```
Comment (lr4d) [inline kartothek/io/eager.py]: you should build indices before you `store_dataframes`, as you require the data to be present at the metapartition (it is gone once you have stored it)

Comment (aniruddhgoteti) [reply kartothek/io/eager.py]: oh ya! Thanks

**Comment by lr4d** [comment]
Looks good but we should add some tests. I will try to get this done for `store_dataset_as_dataframes` which is currently not using the generic tests because of its broken interface. See if you can do this for the other functions

**Comment by lr4d** [comment]
Will have to wait until #126 is merged to get `store_dataset_as_dataframes` conforming to generic tests

File: kartothek/io/dask/delayed.py [hunk 7]
```diff
@@ -156,6 +156,7 @@ def merge_datasets_as_delayed(
     match_how="exact",
     label_merger=None,
     metadata_merger=None,
+    secondary_indices=None,
```
Comment (fjetter) [inline kartothek/io/dask/delayed.py]: Actually, I thought about this pipeline again and I believe we don *not* want to allow index building here. Index building should only be supported for pipelines with immediate write which is not the case here. Sorry for the confusion. Can you please remove this again?

Comment (aniruddhgoteti) [reply kartothek/io/dask/delayed.py]: Sure

**Pull request merged**